src/S3Persistor.js

```diff
@@ -24,343 +23,328 @@ const {
+    }
+    try {
+      await this._getClientForBucket(bucketName).copyObject(params).promise()
+    } catch (err) {
+      throw PersistorHelper.wrapError(
+        err,
+        'failed to copy file in S3',
+        params,
+        WriteError
+      )
+    }
   }
-}
 
-async function checkIfObjectExists(bucketName, key) {
-  try {
-    await getObjectSize(bucketName, key)
-    return true
-  } catch (err) {
-    if (err instanceof NotFoundError) {
-      return false
+  async checkIfObjectExists(bucketName, key) {
+    try {
+      await this.getObjectSize(bucketName, key)
+      return true
+    } catch (err) {
+      if (err instanceof NotFoundError) {
+        return false
+      }
+      throw PersistorHelper.wrapError(
+        err,
+        'error checking whether S3 object exists',
+        { bucketName, key },
+        ReadError
+      )
     }
-    throw PersistorHelper.wrapError(
-      err,
-      'error checking whether S3 object exists',
-      { bucketName, key },
-      ReadError
-    )
   }
-}
 
-async function directorySize(bucketName, key) {
-  try {
-    const response = await _getClientForBucket(bucketName)
-      .listObjects({ Bucket: bucketName, Prefix: key })
-      .promise()
-
-    return response.Contents.reduce((acc, item) => item.Size + acc, 0)
-  } catch (err) {
-    throw PersistorHelper.wrapError(
-      err,
-      'error getting directory size in S3',
-      { bucketName, key },
-      ReadError
-    )
+  async directorySize(bucketName, key) {
+    try {
+      const response = await this._getClientForBucket(bucketName)
+        .listObjects({ Bucket: bucketName, Prefix: key })
+        .promise()
+
+      return response.Contents.reduce((acc, item) => item.Size + acc, 0)
+    } catch (err) {
+      throw PersistorHelper.wrapError(
+        err,
+        'error getting directory size in S3',
+        { bucketName, key },
+        ReadError
+      )
+    }
   }
-}
 
-const _clients = new Map()
-let _defaultClient
+  _getClientForBucket(bucket) {
+    if (this._clients[bucket]) {
+      return this._clients[bucket]
+    }
 
-function _getClientForBucket(bucket) {
-  if (_clients[bucket]) {
-    return _clients[bucket]
-  }
+    if (this.settings.bucketCreds && this.settings.bucketCreds[bucket]) {
+      this._clients[bucket] = new S3(
+        this._buildClientOptions(this.settings.bucketCreds[bucket])
+      )
+      return this._clients[bucket]
+    }
 
-  if (Settings.s3BucketCreds && Settings.s3BucketCreds[bucket]) {
-    _clients[bucket] = new S3(
-      _buildClientOptions(Settings.s3BucketCreds[bucket])
-    )
-    return _clients[bucket]
-  }
+    // no specific credentials for the bucket
+    if (this._defaultClient) {
+      return this._defaultClient
+    }
 
-  // no specific credentials for the bucket
-  if (_defaultClient) {
-    return _defaultClient
-  }
+    if (this.settings.key) {
+      this._defaultClient = new S3(this._buildClientOptions())
+      return this._defaultClient
```

Comment:
Feel free to keep it, but I think this mechanism is less necessary now. You can instantiate separate persistors with different credentials instead. Also, there's an asymmetry because the GCS persistor doesn't support that option.

Reply:
Yes, good point - I think there's a certain amount of faff regarding caching of client objects that isn't necessary here... I don't think they're expensive to build. This can be tidied up.